### PR TITLE
add Identifiable interface - used by Registry and Provider

### PR DESCRIPTION
**Pull request by maggu2810**
Related to: https://github.com/eclipse/smarthome/pull/3670#discussion_r123556157

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/Identifiable.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.core.common.registry;
+
+/**
+ * Interface for classes that instances provide a system wide unique identifier.
+ *
+ * <p>
+ * The requirement "unique" should be seen at least unique per respective type.
+ * Two unrelated stuff that is not hold at the same place every could perhaps return the same identifier.
```
Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/Identifiable.java]: In this sentence something got scrambled

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/Identifiable.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * @author Markus Rathgeb - Initial contribution
+ */
+public interface Identifiable<T> {
+
+    /**
+     * Get the unique identifier.
+     *
+     * @return the unique identifier
+     */
+    T getUID();
```
Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/Identifiable.java]: As we _have_ a `UID` class but are not restricted to it and uniqueness is relative, wouldn't `getId()` more appropriate?

Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/Identifiable.java]: As written in the references issue, most of the respective classes already contain a method `getUid()`. The only one using `getId()` has been the link type.
Rules, tempalte, modules, items, things, ... (IIRC) using `getUID()`.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/Identifiable.java (hunk 1 above)
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/Identifiable.java]: I considered this yesterday, too -- but the PC has already been shut down :wink:

**Comment by maggu2810** [comment]
> Just wondering: Do you already know if this is going to be binary compatible with existing bindings? I.e. do they need to be recompiled?

As long as no changed method is used (WRT the changes to the link) I don't think it should matter.
But I haven't tested it, but recompiled the whole tree.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/Identifiable.java (hunk 2 above)
Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/registry/Identifiable.java]: 👍  very good point, indeed 😒

**Comment by maggu2810** [comment]
What about this one https://github.com/eclipse/smarthome/pull/3762#discussion_r124725109

**Comment by kaikreuzer** [comment]
Oops, sorry, that slipped my attention when merging... Might I ask you for a separate PR for it? 🤕

**Comment by sjka** [comment]
Added the APIBreaking label, as the Registry`s signatures are not binary compatible:

```java
o.e.sh.core.common.registry.Registry.update(java.lang.Object)
```

became

```java
o.e.sh.core.common.registry.Registry.update(o.e.sh.core.common.registry.Identifiable)
```

**Pull request merged**